### DISCUSSION TITLE
TMC2209 and linear advance, question for modders...

### DISCUSSION THREAD

**Discussion by Reggi3** [Jyers/Marlin#927] [Q&A] [answered]
Hi all, I got some interesting information regarding the TMC2209 driver chips on 4.2.2 boards and I was wondering if any of you that have done the mod on a TMC2209 based board could let me know your experiences, how you performed the mod (spread pin, full uart, extruder only uart) and how well the chip behaves with the mod? 

The information I received from !Doman on discord is to stop ALL LOW tmc2209 connection errors, add the following to configuration_adv.h  as per !Doman's instructions.  I do not have this board, so I can't test anything myself but please let us know if this makes a difference to how your TMC drivers behave!!

```
#define  X_SLAVE_ADDRESS 3
#define  Y_SLAVE_ADDRESS 3
#define  Z_SLAVE_ADDRESS 3
#define E0_SLAVE_ADDRESS 3

```
Here is the information:

![image](https://user-images.githubusercontent.com/432746/119995321-5fb39800-bfc5-11eb-9917-e56bcf41e4a7.png)

![image](https://user-images.githubusercontent.com/432746/119994923-fb90d400-bfc4-11eb-89ec-af68d5f93f59.png)

As previously said, if you try this, please let us know how it behaves for you :-)

**Comment by Dom4n** [comment, marked as answer]
Hello, this is !Doman from discord, just wanted to add few things.

Additional change not shown above in `Configuration.h`:
```
#define X_DRIVER_TYPE TMC2209
#define Y_DRIVER_TYPE TMC2209
#define Z_DRIVER_TYPE TMC2209
#define E0_DRIVER_TYPE TMC2209
```
Set driver type to TMC2209 instead of TMC2208. It should work with 2208 too, but if we have this option then why not :+1: 

## Higher resolution images aka why this is needed.
**TMC2209 Datasheet**  
Board 4.2.2 "B" has pins `MS1` `MS2` pulled high.
![image](https://user-images.githubusercontent.com/10115274/119998248-d1ddaa00-bfd0-11eb-839f-fbc33951db8a.png)

**TMC2208 and TMC2225**
![image](https://user-images.githubusercontent.com/10115274/119998342-e5891080-bfd0-11eb-8ad4-e3d5e8ab3776.png)
![image](https://user-images.githubusercontent.com/10115274/119998376-ec178800-bfd0-11eb-94e7-7e38d390d8da.png)

## change working mode
Now switching between StealthChop and SpreadCycle is simple with one command: `M569 S0/1 X Y Z E`

```
Send: M569 S0 X Y Z E
Recv: ok
[...]
Send: M569
Recv: X driver mode:	spreadCycle
Recv: Y driver mode:	spreadCycle
Recv: Z driver mode:	spreadCycle
Recv: E driver mode:	spreadCycle
Recv: ok
[...]
Send: M569 S1 X Y Z E
Recv: ok
[...]
Send: M569
Recv: X driver mode:	stealthChop
Recv: Y driver mode:	stealthChop
Recv: Z driver mode:	stealthChop
Recv: E driver mode:	stealthChop
Recv: ok
```

## hybrid mode
Get best from both worlds and use hybrid mode!
In `Configuration_adv.h` change few values:
```
  #define HYBRID_THRESHOLD

  #define X_HYBRID_THRESHOLD     20  // [mm/s]
  #define Y_HYBRID_THRESHOLD     20
  #define Z_HYBRID_THRESHOLD       3
  #define E0_HYBRID_THRESHOLD     30
```
I wanted to have silent operation when homing or bed level tilting, but when printing with higher speeds (>100mm/s) spreadCycle is better because motors are much cooler and have more torque.

For this to work set driver mode to stealthChop, when threshold is above given value motors will switch to spreadCycle automatically.
